lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -1037,9 +1037,7 @@ def authorized_new(type, value, context)
         end
 
         def lazy?(object)
-          @lazy_cache.fetch(object.class) {
-            @lazy_cache[object.class] = @schema.lazy?(object)
-          }
+          object.graphql_lazy?
```

Comment:
Is there any overhead on calling `object.respond_to?(:graphql_lazy?) && object.graphql_lazy?` over doing the monkey patch in all objects?

Reply:
I _assume_ so, since I often see `Kernel#respond_to?` near the top of Stackprof output (for example: https://github.com/rmosolgo/graphql-ruby/pull/4436). On master, `Kernel#respond_to?` is about 1% of total execution time.

Reply:
To confirm, I tried adding `respond_to?` in bc7e7be6d and ran `bundle exec bench:profile_large_result`, and it gave:

```
Calculating -------------------------------------
Querying for 1000 objects
                          4.457  (± 0.0%) i/s -     23.000  in   5.184884s
==================================
  Mode: wall(1)
  Samples: 674069 (72.81% miss rate)
  GC: 5712 (0.85%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     68566  (10.2%)       68566  (10.2%)     Kernel#respond_to?
     43020   (6.4%)       43020   (6.4%)     Kernel#hash
     28560   (4.2%)       28560   (4.2%)     GraphQL::Execution::Interpreter::Runtime#dead_result?
     26825   (4.0%)       26825   (4.0%)     Thread.current
     22224   (3.3%)       22224   (3.3%)     GraphQL::Schema.lazy_methods
     22211   (3.3%)       22211   (3.3%)     NilClass#===
```

Reply:
Unfortunately `respond_to?` doesn't benefit from method caches, so yes, this is faster. Not sure it's a great idea to add a method to `Object` though (and returning a BasicObject may break it).

Another thing you could try, is to set that hash to compare by identity https://docs.ruby-lang.org/en/master/Hash.html#method-i-compare_by_identity

That will save the call to `#hash`. With that you may have similarish performance, especially since this `#graphql_lazy?` is likely to be very polymorphic, and the interpreter can only cache one class.

Reply:
Trying it out over here: https://github.com/rmosolgo/graphql-ruby/pull/4443

It did save a lot of time in `Kernel#hash` (and could save _more_ as mentioned in the TODOs on that PR), but didn't save any time from the `Kernel#class` time.

Reply:
> but didn't save any time from the `Kernel#class` time.

Yes as expected. But the hope is that the difference is now small enough that it doesn't really justify a core extension into Object.

Reply:
I _did_ get a ~10% speedup on that PR by using `compare_by_identity` and re-working some `lazy?` checks to use that cache instead of the less-optimized schema-level cache. This would probably _still_ give another big improvement, since `Kernel#class` is now 10% of stackprof time on my microbenchmark, but it certainly worked out well over there.